### ISSUE TITLE
Team review of template

### ISSUE THREAD

**Issue by lboxell** [gslab-econ/template#19]
Evaluation of template in parallel with gslab_scons module in https://github.com/gslab-econ/gslab_python/issues/16. 

Both small template (`master`) and large template (`large_template`). 

Leave comments in task and I'll aggregate later.

**Comment by jmshapir** [comment]
Am going through small template first.

Quick question about how paths are being handled. How does [`analysis.do`](https://github.com/gslab-econ/template/blob/master/source/analysis/analysis.do) know that `output/data/data.txt` is relative to root rather than relative to the location of analysis.do? Is this something we've implemented via the builder?

**Comment by lboxell** [comment]
@jmshapir,

When scons runs build steps, it should run them from the root (wherever the SConstruct is located). So all paths within code scripts should be relative to the root. For SConscript files, paths are relative to the location of the SConscript file (unless the `#` flag is used to denote the root directory). This is my understanding of how scons behaves by default -- nothing that we've modified.

**Comment by arosenbe** [comment]
@lboxell, my comments are below. I'll post a new comment if I find anything else. 

### Both Templates

  *  It's misleading to have the small template in `master` and the large template in `large_template` since neither of them dominates 

  *  Readmes
    *  The grammar needs to be cleaned up in the prerequisites and in the readme to the linked gslab_latex repository. All the writing could probably do with a polish. 
    *  The "additional information" points at the bottom of the prerequisites look like finer points about the Metropolis Beamer Theme, but they aren't. They are also poorly formatted.
    *  Clean up the wiki formatting
    *  Remove the links to scons resources. These are in the RA Manual. We could link to that instead.
    *  We should try to standardize the example code in the two templates as much as possible.

### Small Template

  *  We should bring our naming conventions in line with the large template. I'm specifically thinking about about renaming `output` as `build`.

### Large Template

  *  Readme
    *  Fix grammar in `cache` and `release` protocol
    *  The `cache` and `release` protocol is unique to GSLab and probably not appropriate for a public repo.
    *  Does this template still not support windows?
    *  The **To run** instructions are out of date

**Comment by gentzkow** [comment]
Initial thought: I would vote that we get rid of the small template entirely. Originally, I had thought the large template would require a much more elaborate structure. But we have managed to simplify it using cache etc. to the point that a lot of the complexity is hidden and I think we can get by with just one.

At some point, I'm going to think about whether we might want to have a non-scons version for use with co-authors who aren't going to want to learn the new tool. But that can wait for another day.

**Comment by jmshapir** [comment]
Question for everyone: should we add ```ondeck.lyx``` and ```text.lyx``` to the [```/paper```](https://github.com/gslab-econ/template/tree/master/source/paper) sub?

If the goal is just to illustrate repository structure then we don't need those. But if we want this to serve as the template for new projects then it will be helpful to have them.

**Comment by jmshapir** [comment]
Another question: Do we want to develop some helper tools that automate detection of source files? 

Use case: I write a slide deck that uses 100 different politext plots. I could itemize all of them in the ```SConscript``` in ```/slides``` but that is in a sense redundant with the information in the LyX file. We could have the structure

```
plots = `detect_plots('#source/talk/slides.lyx')
target = '#output/talk/slides.pdf'
source = ['#source/talk/slides.lyx',
		  plots]'
```

**Comment by jmshapir** [comment]
More radical question: what about adding a high-level ```/tables``` sub and including each table as its own file? We could then do table filling separately from draft compiling. Dependency management would be likely be simpler because once tables are filled we'd just include tables as LyX includes, so some of the messy dependencies would not show up in the ```/paper``` sub.

A downside is it would require more navigation to get to our table notes and stuff.

Again I'm not sure if this is in scope for this review since it relates to our overall practices rather than this minimal template. But it's something I've been thinking about and wanted to surface.

**Comment by jmshapir** [comment]
I'm looking now at "large." A couple questions that the team can answer to help me navigate:
- Are we no longer using the gdrive cli? 
- If we have some large raw data that we are interacting with, will that live someplace like ```/Users/%s/Google Drive/raw```?

**Comment by lboxell** [comment]
@jmshapir,

[This](https://github.com/gslab-econ/congress_text/blob/refactor_code/source/data/SConscript) is how we are currently dealing with raw data in congress_text. I think our current implementation of the template and module is flexible enough that we can vary the approach based on the actual size of the data. For congress_text, the data is small enough that we can download it and cache it without many issues. If we were dealing with 100's GB, then we could store it in a single place like  `/Users/%s/Google Drive/raw` or some network drive and just symlink to it.

**Comment by gentzkow** [comment]
@jmshapir, @lboxell: Just a reminder that we also have a long list of suggestions for the template on the [admin outline](https://github.com/gslab-econ/admin/wiki/Outline). I would vote that we make it part of this task to review these and decide which we plan to implement.

**Comment by gentzkow** [comment]
To @jmshapir's question above.

* Yes to ondeck.lyx and text.lyx
* Yes to writing a [scanner](http://www.scons.org/doc/2.0.1/HTML/scons-user/c3833.html) for lyx files
* Yes to `/tables` sub. We could even create a `/figures` sub as well with a build step that just copies the subset of figures we plan to include in paper / talk from the various locations where they were built to `/figures`. For big projects like politext it would be very nice to have all the core figures in one place and a map of where they came from rather than having elaborate paths to the original locations baked directly into the Lyx file the way we do now.

**Comment by gentzkow** [comment]
First set of comments. These are things I think we can implement immediately.

### Readme / documentation
* Lab-specific protocol info should be moved to RA manual
* Google drive should be dropped as a requirement; assume generic storage 
* Discussion of requirements could be pared down / simplified further

### Setup
* I had to explicitly install the Python `requests` library to run the large template. Looks like this is something gslab_scons depends on. Is there a way to specify things so this (and any other dependencies) get installed automatically when we install gslab_python?
* I would like to define an explicit `setup.py` (or equivalent step within scons) which verifies that all prerequisites are satisfied. The goal should be that we would trap most errors at this stage so if a new user passes the setup step the repository runs without error with very high probability. It could also be part of this step to check out externals from SVN + create sim links to external raw data / code / etc.

### Other
* I wonder if we don't want to drop sconstruct.log and sconscript.log entirely. Since these only record the last run of scons, they are not all that useful; e.g., if my practice is always to run scons a last time to make sure the build is "clean" then these will be totally uninformative
* Per my earlier comment, I'd vote we combine the small and large templates; if we don't, we should update the large template to include all the elements (`/paper`, `/talk`, etc.)
* The SConscript in /data/ in the lg template should use build_r not env.Command()
* What about renaming the builders we define in SConstruct `BuildLyx`, `BuildStata`, etc. rather than `Lyx`, `Stata`, etc.?
* We should think about the templates we want R and Stata scripts to follow and implement those here. For Stata, this will mean revisiting our gslab-specific tools like `preliminaries` -- whether we want to keep using them and also how we want to install them.
* I don't think we want very short scripts like the current build.R to use the

```
main <- function() {
}
main()
```

structure.

**Comment by jmshapir** [comment]
@gentzkow two things:
1. On having a /figures sub, I agree with that but would propose something different, which is that (in parallel with tables) we have a LyX file for each figure that houses the figure and its notes. Then these are called as individual child documents in papers.lyx (and other draft documents). This way dependencies/paths in papers.lyx are very simple and clean.
2. I like the ```sconstruct.log```. Even though often it will not say much it will serve a good way to test if a commit was valid.

**Comment by gentzkow** [comment]
Second set of comments. These are places where I don't have a firm idea what the right decision is and I think we should discuss.

### Handling raw data / externals
* I think we should put all such files (either the files themselves or symlinks to them in a directory called `/build/external`
* I would vote that we note cache these external files; general principle should be we never store duplicate copies of large files if we can avoid it
* For files that are expected to be stored locally (e.g., raw data downloaded via Google Drive, Dropbox, etc.), I would vote that the user be able to specify the path to the local directory and the build step creates a symlink inside `/build/external`. We'll need to find a way to do this so that the paths I enter locally remain in my local repo but don't get committed so that they will overwrite the values other users have set.
* For files we are exporting from SVN or Github, we could either place them directly in `/build/external` or require that the user check them out manually to a separate local location and then point to the appropriate path.
* Do we want to have a way of recording the state of the external directory at the point the code runs? E.g., print a manifest of all files with their sizes & time stamps? The scons .dblite file will already record the hashes of any that we depend on, but it's a fair bit of work to reverse engineer this and it might be nice to have a simple record.

### What gets committed
* We need to resolve what if any output files we want to commit to git / git-lfs. Options include:
    - Nothing gets committed; we release frequently enough (e.g., whenever an issue is complete) that in most cases where we want to look at the current or past version of the paper, etc. we can find it in a release. In the rare cases where we need an intermediate version we have it in the cache or at worst we can rebuild it from code.
    - We have separate directories for committed and uncommitted output. We could call these `/output` and `/release` or something similar
    - We commit all or some of the `/build` directory
* If we could make it work I like the first option. In working with git-lfs I've found it creates an extra layer that can cause bugs / problems -- e.g., my desktop GitHub client often crashes when trying to pull / push large numbers of binary files via lfs. But I also don't want us to lose the ability to easily browse current / past files.

### Release protocol
* I wonder if we don't want to take this out of scons and just have a release.py script. This seems like the kind of thing you want to be as simple & transparent as possible, and it doesn't require dependency tracking.
* For big projects like politext, what gets released? Just the paper & slides? Those plus the tables & figures? All of the analysis output? Any of the intermediate data? Do we want to have a protocol for releasing different subsets of these at different times?

### Issues
* Where do issue markdown files fit in the template?
* I'd like to have a protocol that allows issue deliverables to be compiled pdfs including tables, figures, and other output, with clear provenance for where the output files came from. Simplest way would seem to be that we can release from an issue branch and in this case the released files are just the issue deliverables. Should think about how to make this work.

### Paths
* We still need to resolve the paths issue that we addressed in politext using paths.txt. What I don't think we want is to have specific paths like /build/analysis/model_runs/no_demographics/ etc. hard-coded into scripts. I should be able to change such paths easily, and also to look and see easily where a directory is getting input and placing output.
* I'm find sticking to paths.txt as a solution. Another possibility is to use the fact that scons actually runs a copy of the code in the /build/ directory (see discussion [here](http://scons.org/doc/1.2.0/HTML/scons-user/c3271.html)) and so write the code as if we are building everything locally inside the source directory. Not sure yet whether this would work and/or be desirable.
* Should we be using the scons Depends() functionality for input files (e.g., dataset that a Stata script requires)?

**Comment by gentzkow** [comment]
OK. I'm fine with your view on both counts.

**Comment by lboxell** [comment]
@gentzkow,

My comments on your comments:

### Handling raw data / externals
 - Storing in build/externals makes sense.
 - I think for files where we don't have a good sense of a revision number, it would be good to have some way of determining the state of the file when scons was run beyond the .dblite file, though I'm not sure what this would look like.
 - I'd be hesitant to have manual download/sync steps for raw data. I think this would be hard to coordinate if there are different versions of raw data floating around and cause issues with dependency tracking. In general, I don't think there is a one-size-fits-all approach to this aspect of the workflow. 

### What gets committed
 - I'd vote to have an 'output' folder that gets versioned in GitHub. One case study is the congress_text repo, we are now starting to have these 'test' output files that print comparisons and performance measurements. When we version control these, it makes understanding how changes in the code impacted output easier if these files automatically tracked and diffed by git. Otherwise, we have to do manual diffing (or manually diff with some automatic differ). 

### Release protocol
* Integrating the release step with scons helps ensure that everything is up to date and built before being released, though I agree having a release.py script could be simpler/clearer for people. (The release.py could probably do a `scons --dry-run` to check and see if everything is built too.) Would the release.py live in the root of the template?
* Regardless of the above point, I think it would be good if our release protocol was flexible enough to release different subsets for different milestones. So, for a paper submission, we probably want to release as much as possible, but for a small change to the paper, we probably don't want to re-release everything. 

### Issues
* I believe our current implementation of the scons release mode will allow releasing from a branch instead of master.

### Paths
* Paths.txt was working well for the most part in politext. I think we'll want to refactor it though so it's a little easier to use when coding interactively. 
* My understanding is that the Depends() function is essentially the same as specifying a 'source' file in a builder.

**Comment by jmshapir** [comment]
In terms of what gets committed, I'd like to preserve the ability to easily browse plots and other light output in the browser. I realize there's some overhead with using git-lfs to handle binaries but I think we've benefited a lot from being able to link to intermediate versions of plots and things in politext. But I'm open to discussion as I understand there are other ways to organize that (e.g. posting lots of screenshots in issues, etc.)

**Comment by gentzkow** [comment]
Thanks guys.

I'd like to discuss some of these things face to face. I could do that w/
just Levi next week, or we could find a time to talk 3-way.

**Comment by jmshapir** [comment]
@gentzkow et al. I'm happy to do a lab skype to talk about this stuff. I have a standing call with the whole team Thursdays at 2:30 ET that we could use. 

An alternative would be to have you guys get carry out for "Noodles" and for us to skype during that time to discuss.

**Comment by yuchuan2016** [comment]
Hi @gentzkow,

A quick reply to your question in [setup](https://github.com/gslab-econ/template/issues/19#issuecomment-259769756). We could add an argument `install_requires` in the setup.py in gslab_python to specify what it minimally needs to run correctly. Then when the project is installed by pip, all dependencies specified here can be installed as requirements. See [here](https://packaging.python.org/requirements/#install-requires).

**Comment by jmshapir** [comment]
@gentzkow if you want we could use my skype at 2:30ET Thursday to resolve the remaining issue with paths.txt etc. that we didn't cover in today's call.

**Comment by gentzkow** [comment]
Sure. Good idea.

**Comment by lboxell** [comment]
@jmshapir @gentzkow @arosenbe @stanfordquan @yuchuan2016 @M-R-Sullivan:

## Consensus from today's meeting: 

### raw data & externals
 * Store in build/externals, but do NOT cache. 
 * 3 options to populate data/externals (in decreasing robustness):
 	* Download from gdrive/etc using some revision number
 	* Point to a synced version of gdrive/etc and (optionally) check that it is up-to-date
 		* Make some checker to verify the directory is up-to-date.
 	* Point to a local directory and the user must verify that it is the correct version
 * Template comes with some 'settings' file that each user is able to modify and specify where the 'cache' folder, any local externals directories, etc files might be located. This file is not versioned.
 	* This 'settings' folder should be printed to a log that is versioned.
 * Should create some logging tool that prints the entire 'state' of a repo at the time of a run.
  

 ### What gets committed and releasing
  * Have a /release/ directory that stores the 'output' of repo.
  * Some items in /release/ get versioned in git, others do not.
  		* Small items such as plots, figures, tables, etc. should be versioned.
  * Every time you merge into a 'master' branch, the source and release files should match, but issue branches are free to follow other workflows assuming a lot of storage is not being consumed from commits, with two primary options:
  		* scons mode=dev (where /release/ is not populated)
  		* OR, just don't add /release/ when you commit in an issue branch.
  * Whenever you make a release, the entire /release/ folder gets added to Gdrive and we have a pointer in GitHub to this location. We attach no other binaries to the github release besides this pointer file.
  * Create a tool that zips and stores a 'snapshot' of a repo that compresses all files tracked by the 'state of the repo' logger.
  * Have a separate `release.py` script that checks to ensure scons run is up to date and makes the release. This could either sit in the template repo or in gslab_python depending on how much customization is needed for each repo. 


### Issues
  * Root level /issues/ folder in each issue_branch, which is versioned in git.
  * All figures and tables should be able to be traced back to how they were created, but using scons to populate is not necessary.
  * This folder does NOT get merged to master.
  * A revision stable link should be posted at the end of an issue's comment thread to the final deliverable. 

### Other
  * Pare down the RA manual's discussion of template and move as much as possible to the template to be self-documenting (but don't move lab-specific details)
  * gslab_stata should be installed to each local machine as a part of the computer build process and listed as a pre-req for any template that uses stata. 



## Remaining questions (from this task and related tasks and things I thought of post-meeting):
  * What about small raw data that is repo specific? Should we have a /raw/ or source/raw/ folder that allows such inputs and just version them in git?
  * Do we want a hard rule (or some guideline) on the size of files that can be versioned in release?
  * Paths.txt
  * Passing command line arguments to scons builders
  * Log files for each build step and versioning them. 
  * Use of env.INPUT (?) in scons builders

**Comment by jmshapir** [comment]
@lboxell a few additions on raw data & externals:
- We also decided the template would have a committed 'externals' file that always points to the authoritative location/version of the file in the data store. By default unless overridden by user settings, data will always be fetched from the location specified in externals. User-defined settings can override this by, say, directing fetcher to create symlink to local copy.
- The 'state' of the repo we want to print is basically a manifest of files and file properties. We want something that involves no computation.

**Comment by lboxell** [comment]
I should have tasks open to address all of the above points, except "Create a tool that zips and stores a 'snapshot' of a repo that compresses all files tracked by the 'state of the repo' logger," which I added to the admin outline.

**Comment by lboxell** [comment]
@gentzkow @jmshapir 



### Paths.txt

General discussion on paths.txt for @gentzkow and @jmshapir to provide input on based on skype + RA discussion. There are two related decisions: how to load paths and where to store the paths.

How to load paths:
 - Environment variables
 	- Would still need to load from some file to set the environment variables and then each code script would have to read the environment variables.
 - Read directly from .txt file
 	- Code and Scons just read directly from .txt file and creates dictionary that we call paths from. Avoids namespace conflicts and potentially lingering environment variables.

 Proposal: Read directly from .txt file. Build helper function for stata, R, and python to do this that lives in each language's gslab_* lib. 


Two options for where these paths.txt files live.
 - One master paths.txt at the root of the directory.
 	- All objects would have distinct names. 
 	- Could get clunky and long.
 	- Could use environment variables.
 	- Change everything in one place. 
 - Individual paths.txt for each SConscript file. 
 	- Using environment variables would cause namespace conflicts with scons. 
 	- Could use the same name for conceptually similar objects (e.g., 'OUTPUT_DIR').
 	- Changes to directory structure would require more changes.
 	- Directory re-organization would require changing each individual paths.txt (probably could be accomplished with grep type command line argument)

 Proposal: Individual paths.txt for each SConscript. These paths.txt are not specified as sources to build steps because the paths already take care of this (if you change the paths in paths.txt, I think scons will force a re-run. Not really sure about this, but we can test it out as we build the tools.)

Thoughts?

### Other questions for @gentzkow (Jesse deferred)
  - Use of env.INPUT?
  - Use of Depends vs specifying source in build steps? Is it just a difference in preferred style?
  - Where should we store small raw input that is repo-specific? In source/raw/? In /raw/?
  - Do we want a hard rule (or some guideline) on the size of files that can be versioned on git in the release?

  - For each build step, we create a log file and we specify this as a target and install it in release. (We discussed this a bit with Jesse, who likes the idea of versioning this information, but we didn't discuss implementation details.)

### Things for RAs to spend some more time on 
  - Command line arguments in scons builders

**Comment by jmshapir** [comment]
@lboxell @gentzkow I don't agree with @lboxell's proposals for paths.txt.

Instead I propose:
- A file called ```aliases.txt``` lives at the root alongside the ```SConstruct```.
- The ```SConstruct``` slurps aliases.txt and converts all aliases into environment variables.

Notes:
- My understanding is that scons can create environment variables that persist during the build and then vanish thereafter. This means we would not need to have individual scripts "read" environment variables. They would just call them the way they normally access OS environment information.
- My vision would be that aliases.txt would mostly be paths to resources but would not have to be. It would be a way to have soft-coded representations of anything. It could even be key numbers (e.g. number of bootstrap runs) that we want to persist across all runs.
- This would create a single source of truth so that if we, say, reorganize paths, we only change one file. I don't think it would require writing any new helpers except for something that helps scons load these as environment variables, which I think would be simple.

**Comment by gentzkow** [comment]
I am in favor of .txt file rather than environment variables.

Main reason: I think it's important to be able to run code interactively /
ad hoc rather than only inside an scons run

Other reasons:
(1) Each script would need special code to either load txt file or read
environment variables, so cost in this respect is equivalent; may even be
higher for environment variables because not all software makes it easy to
load these and in some cases (stata?) we'd need different code depending on
the OS.
(2) Environment variables seem likely to introduce more scope for bugs /
errors / cross-platform compatibility issues

I don't feel strongly about root vs. in individual directories, but I tend
to favor the latter because a single file would get unwieldy for a project
on the scale of politext.

Happy to discuss.

I'll respond to the other queries shortly.

**Comment by jmshapir** [comment]
@gentzkow I don't think I follow re: environment variables. In Stata can't I say 

local name : env ENVOBJECT

and immediately call the name of whatever is in ENVOBJECT, regardless of my OS? I would have thought basically all tools allow direct access to environment variables in something like this fashion. In R it is called sys.getenv(), etc.

**Comment by gentzkow** [comment]
@jmshapir Ah, hadn't realized that; thought we'd have to do it with a call to the command line. So you can scratch that second part of (1).

Still the main thing seems to me to be the ability to run code interactively.

What do you see as the advantages of environment vars?

**Comment by jmshapir** [comment]
@gentzkow here is how I have been thinking about it.

At a high level:
- The problem we are trying to solve is to have some information persist during the build and be accessible to all scripts. Our build framework intends environment variables as a solution to that problem. I'm in favor of trying to use the tool as intended.
- To me this is broader than paths. We often want to have constants floating around, say for example the minimum number of uses of a phrase for it to appear in the sample. This is the problem we have solved in the past with ```loadglob``` and its friends. If we solve the paths problem I would ideally like to use the same solution for any constants or any other values that we want to be shared across scripts in the build and that we only want to have to change in one place.

At a lower level:
- As always I would like to avoid redundancy as much as possible. So whatever we do I am in favor of specifying these aliases in only one place if we can make that work.
- The issue of being able to use the code interactively seems minor to me. For example we need to write some kind of python script to slurp in values and set them as environment variables. We could have a variant of the same script that populates system (as opposed to build) environment variables. We'd have to be careful with name conflicts (e.g., we wouldn't want to have something called PATH or USERNAME) but we probably want to be careful about name conflict in any case.

**Comment by jmshapir** [comment]
@gentzkow following our call yesterday I spent some time looking around. I don't think I have found anything decisive (hopefully you have) but wanted to report in.

* von Gaudecker, who is working in Waf, [advocates](http://hmgaudecker.github.io/econ-project-templates/organisation.html) storing project paths in the main wscript and then resolving them into absolute paths in the scripts below. It seems like this would have some of the same pros and cons we've already talked about.
* The other advocate of make frameworks for research I know if is Hyndman, but his [github page](https://github.com/robjhyndman?tab=repositories) mainly has R packages and so doesn't illustrate interacting with data.
* I took a look at a few github repos that could be useful references, such as [congress-legislators](https://github.com/unitedstates/congress-legislators), [git-lfs](https://github.com/git-lfs/git-lfs), and a [library built by civis](https://github.com/civisanalytics/postgis). These are not repositories that do data analysis so they are not a great comparison. But what I found interesting is that their path structure is very shallow.
* Horton has a [public repo](https://github.com/johnjosephhorton/sharing) for replicating his sharing economy paper with Zeckhauser. It looks like he handles data by creating a loadable R package that is executed to create the necessary data frame. This works well here because the data come from two small files and all the analysis is in R.

**Comment by gentzkow** [comment]
@jmshapir: Thanks.

The first piece of this I have considered is defining a search path for input files that will be opened / loaded by scripts. My overwhelming sense is that this is *not* common practice. Consistently, the standard seems to be to define search paths for code modules / packages / libraries but *not* to do so for data or other input files. E.g.:

* Python has a carefully constructed path structure for code libraries, but not for input files. There are several stackexchange threads where people ask about this and are told it doesn't work like that (e.g., [here](http://stackoverflow.com/questions/12391709/open-cant-find-file-given-path-relative-to-pythonpath)).

* R uses [libpaths](http://www.astrostatistics.psu.edu/su07/R/html/base/html/libPaths.html) as a search path for libraries but does not define anything analogous for data/input files.

* Stata uses [adopath](http://www.stata.com/manuals13/psysdir.pdf) as a search path for .ado files but does not define anything analogous for data/input files.

**Comment by gentzkow** [comment]
@jmshapir: Second piece is using text files vs environment variables. We have talked about two distinct use cases: one is defining user-specific variables like location of an external data source which are *not* committed to the repo; the other is defining local paths and other variables that will be the same for all users as *are* committed to the repo.

There is a lot of discussion of this online for the *former* (user-specific) case. Seems like there is an active debate, with one prominent source ("12 Factor App") advocating environment variables and a number of others arguing it's better to use text files. Note that this is all in the context of deploying web apps, but I think the issues are mostly the same.

Arguments for enviro vars:
* [12 factor](https://12factor.net/config)
* [Heroku](https://devcenter.heroku.com/articles/config-vars)

Arguments for text files:
* [Environment Variables and Why You Shouldn't Use Them](https://support.cloud.engineyard.com/hc/en-us/articles/205407508-Environment-Variables-and-Why-You-Shouldn-t-Use-Them)
* [Environment Variable Considered Harmful for your Secrets](http://movingfast.io/articles/environment-variables-considered-harmful/)
* [12 factor app configuration vs leaking environment variables](https://gist.github.com/telent/9742059)

I think a fair reading of this is that either approach is defensible and neither has a clear advantage in terms of being considered best practice.

**Comment by gentzkow** [comment]
For the *second* (non-user-specific) use case, all the references I can find point toward using text files. 

Examples:

* [cross-language constants](http://codereview.stackexchange.com/questions/19953/cross-language-constants)
* [sharing constants across languages](http://stackoverflow.com/questions/3550299/sharing-constants-across-languages)

One point to note is that in all these solutions people are storing constants using a standard data format like [xml](https://en.wikipedia.org/wiki/XML) or [yaml](https://en.wikipedia.org/wiki/YAML). This seems like a very good idea to me, and it would mean we don't need to write custom tools to read the text files, at least for R, Python, etc. (though we would probably want a custom wrapper that would convert the data read in to local / global variables).

**Comment by gentzkow** [comment]
For deciding among xml, json, yaml, etc. there is some discussion [here](http://electronicdesign.com/dev-tools/whats-difference-between-json-xml-and-yaml).

All sources agree that yaml is the simplest, and closest in spirit to Python (since it uses whitespace and returns to define hierarchy).

**Comment by gentzkow** [comment]
@jmshapir, @lboxell: 

Putting all this together, my vote on the paths.txt issue is that we settle on the text file solution with the following modifications / clarifications:

* We apply the same approach to both non-user-specific constants like relative paths and other constants we want to share across code files, as well as user-specific constants like raw data locations or login credentials

* Constants are stored in flat .yaml files; we use existing tools to read yaml format rather than writing our own whenever possible

* File names are constants.yaml for non-user specific and user-config.yaml for user-specific. The user-config.yaml file is not committed to the repo (the template specifies this in .gitignore). We can provide a template for user-config.yaml inside the README.md

* We allow constants.yaml files to exist anywhere there is an sconstruct or sconscript. What goes where is decided to limit redundancy and improve clarity. By default for simple projects we define everything in the top-level constants.yaml file.

(My thinking on the last point is that the added clarity of putting constants deeper in the tree will in some cases outweigh the cost of redundancy, especially given that it will be easy to modify redundant constants throughout the tree using `grep`. But we can revisit once we gain some experience using this structure in large trees.)

**Comment by gentzkow** [comment]
Here are replies to the following queries from @lboxell's comment above.

### Other questions for @gentzkow (Jesse deferred)
  - Use of env.INPUT?
  - Use of Depends vs specifying source in build steps? Is it just a difference in preferred style?
  - Where should we store small raw input that is repo-specific? In source/raw/? In /raw/?
  - Do we want a hard rule (or some guideline) on the size of files that can be versioned on git in the release?

#### env.INPUT
My thought was that we might prefer to separate input files like figures / tables / etc. from source code files. E.g., in the `die` repository I had set up tablefill to look like this:

* SConscript *
inputs = ['#external/tables/DocsSumStats/SummaryStats.txt',
             '#external/tables/HazardModel/hazard_mortality_cox.txt']
source = ['#/source/paper/tables.lyx']
target = ['#output/paper/tables_filled.lyx']
env.Tablefill(target, source, INPUTS=inputs)

* gslab_scons.py *
def build_tables(target, source, env):
    tablefill(input    = ' '.join(env.GetBuildPath(env['INPUTS'])), 
              template = env.GetBuildPath(source[0]), 
              output   = env.GetBuildPath(target[0]))
    return None

This seems clearer and more robust to me than separating the source and input files based on their position in the `source` list.

Of course we'd need to confirm that we can do this in a way that SCons will treat what's input as dependencies and rebuild if they are modified. I believe we can do this using Depends().


#### Depends
We can ignore this comment. I was misremembering what Depends() does in SCons.


#### Small raw data files
I'd vote we put this in `/source/input/`. This should be a legal place to put not only data files but other inputs such as images, etc. I don't feel strongly about "input" vs "raw," though, so I'm fine with the latter if you guys prefer.

#### Size of files that can be versioned
I do think we should impose a cap as a default, though we can decide to break it in special cases. Maybe 2 MB? That should give us room to store PDFs but not let us store anything that's going to impose a significant cost. We may also want to impose a separate cap on the total size of the /release/ directory that can be versioned. Maybe 500 MB?

If we impose a cap, we should think about if there's a way to enforce it with code. One possibility is to write a little python script that walks the release directory skipping things in .gitignore, then outputs a warning if you're over the cap. A second is to use some kind of commit hook in Git. I think I like the warning version better since it's less likely to be a source of bugs and it will be easy to override if we want to.

**Comment by jmshapir** [comment]
@gentzkow your solution [here](https://github.com/gslab-econ/template/issues/19#issuecomment-262019876) works for me. It even looks like [github will render yaml](https://github.com/blog/1647-viewing-yaml-metadata-in-your-documents). I would suggest we call the files global-config.yaml and user-config.yaml for parallelism but I am happy with constants.yaml if you prefer that.

I don't have strong feelings about any of [these items](https://github.com/gslab-econ/template/issues/19#issuecomment-262024738) so am happy to defer to @gentzkow on those.

**Comment by gentzkow** [comment]
Great. I'm happy with global-config.yaml.

**Comment by lboxell** [comment]
@gentzkow,

Sounds good. I updated the related tasks with the additional information.  

I think there is one last issue, do you agree with the following point from my previous comment?

> For each build step, we create a log file and we specify this as a target and install it in release. (We discussed this a bit with Jesse, who likes the idea of versioning this information, but we didn't discuss implementation details.)

**Comment by gentzkow** [comment]
@lboxell: Sorry I missed that. When you say each build step you mean each individual call to a builder in scons? So we'd have one log file associated with each script?

I'm not sure about this. It seems like it might be overkill to require this as a default. E.g., we have to have log files for tablefill, each time we compile a lyx doc, each R script we run, etc. My instinct would have been to do this only in cases where the build step writes information to the log file that is indeed part of the output we intend to release / use downstream (e.g., a .do file that writes regression output to a lot).

There is a separate issue about logging what scons is doing rather than what individual builders are doing. We talked about this a while back and decided it was best not to define these log files as targets. I remain unsure how valuable it is to have these at all, but I'm happy to keep them if you or @jmshapir think they're valuable.

**Comment by lboxell** [comment]
@gentzkow,

I agree doing it for every build step would be overkill. There are two things that I think would be useful to track (and @jmshapir seemed to agree) for certain build steps:
 - the execution time (so we can understand the costs of having to re-run). 
 - if there were any warnings() or other non-exit messages

**Comment by gentzkow** [comment]
I agree. I think the ideal thing would be to have a single log file that
tracks these things and is versioned but not defined as an scons target.

What about defining a log file location in our scons build environment and
then have each of our custom builders write their start and end times to
that log. Trapping errors may require more work; probably need to handle
this separately for each of Stata, R, Lyx, etc. But if we can collect the
errors successfully we can also write those to the log.

If we do this, I'd vote that we drop the separate SConstruct.log.

**Comment by gentzkow** [comment]
(Alternatively, we can see if there is other SCons functionality that will
allow us to log start/end times for individual builders; i.e., how would we
do this if we couldn't touch the code inside the builders themselves?)

**Comment by lboxell** [comment]
@gentzkow,

I don't see a great way to do this with a single log file. For example, if we modify something which causes one build step to be re-run, but not another, how do we update the log file to track the new run time for one build step while not losing the old run time for the other? 

The only way I can think of right now with a single log file is to create something like:

```
build/data/text.txt : 1hr 30 min 4 sec
build/analysis/plot.pdf : .1 sec
```

and then the logger has to first search through the log file to see if a given target already exists and then either replaces this log time or appends a new dictionary pair, where by default, the first target is used for the dictionary name. Or something like that. 

Did you have an alternative solution in mind? 


I'm currently looking into [scons-time](http://www.scons.org/doc/HTML/scons-time-man.html) and this site that uses [scons TIMER=y](http://www.reproducibility.org/RSF/book/rsf/scons/paper_html/node7.html).

**Comment by lboxell** [comment]
The key issue in the above is if we run scons multiple times between commits. If we only run scons once between commits we can always find old run times in version history as @jmshapir suggested in https://github.com/gslab-econ/gslab_python/issues/16#issuecomment-260089367). 


It looks like `scons --debug=time` might do what we want. Though it will only print these run times when it actually runs a build step, so we'd still have to move them to a 'safe' place to commit per the issue noted above. 

Example output:
```
Command execution time: build/congress_members_001.txt: 120.153494 seconds
Command execution time: build: 0.000006 seconds
Command execution time: .: 0.000004 seconds
scons: done building targets.
Total build time: 120.691518 seconds
Total SConscript file execution time: 0.352648 seconds
Total SCons execution time: 0.185366 seconds
Total command execution time: 120.153504 seconds
```

**Comment by gentzkow** [comment]
Hmm. A couple of options.

(1) We log only the builds scons runs, but we append rather than overwrite;
so the log accumulates a record of every run. One can easily search to find
the most recent run of a particular build.

(2) We make this part of the build step, but only do it for slow steps for
which we think we'll actually care about the execution time. Then these
steps would have a log file target just like the Stata scripts.

I think @lboxell and @jmshapir have a clearer sense than I do of the
purpose of these logs, so I'm happy to leave it to you guys to decide. For
myself, I sort of like the simplicity of (1).

**Comment by lboxell** [comment]
@jmshapir,

My thoughts are that (1) might get pretty long, especially during development when things are run multiple times. I'd vote for (2), but I think either option would work. Let me know if you have a different opinion.

**Comment by jmshapir** [comment]
@lboxell I am happy to leave this in your hands as long as what you come up with will work for you and is acceptable to @gentzkow.

I will just give you a sense of what I am trying to achieve.

Most of the time everything works smoothly and it's easy to tell what scripts were executed on a given commit. From time to time weird things will happen and we'll want to go back to the commit history to understand where a given output came from. In such cases it will be helpful to know what scripts we ran between commits. 

Likewise, it would be ideal to have an easy way to know if scripts are throwing errors to the OS so that we can fix those before committing or so that we can understand why a commit looks weird.

If there is a good way to achieve these goals without much additional work and complexity I would be in favor of it.

**Comment by gentzkow** [comment]
My 2 cents given @jmshapir's reply: I'd vote we start with my option (1)
and see how that goes. It's true the log file will get very long, but we
should be able to structure it so it's easy to search / filter (e.g., to
look at the most recent run of each directory), and I doubt it will get so
big as to pose a storage problem. There may be rare cases where having the
full record of every run that took place will be valuable.

If we find it gets too clunky, we could add a script at the end of the
scons run that filters the log file keeping only the most recent run of
each directory.

**Comment by lboxell** [comment]
Sounds good. I updated the relevant task description. 

I think all the discussions from this task have open tasks or were added to the outline, so I think we can close this.
